partition_snapshot_reader.hh

```diff
@@ -106,9 +123,9 @@ class partition_snapshot_flat_reader : public flat_mutation_reader::impl, public
 
                 range_tombstone_list::iterator_range rt_slice = [&] () {
                     if (last_rts) {
-                        return v.partition().row_tombstones().upper_slice(_schema, *last_rts, bound_view::from_range_end(ck_range));
+                        return v.partition().row_tombstones().upper_slice(*_partition_schema, *last_rts, bound_view::from_range_end(ck_range));
                     } else {
-                        return v.partition().row_tombstones().slice(_schema, ck_range);
+                        return v.partition().row_tombstones().slice(*_partition_schema, ck_range);
```

Comment:
So _range_tombstones will hold ranges sorted by the original (non-reversed) order (_heap_cmp). Isn't that wrong? If we have two ranges: ([2, 3], [5, 6]) and ([0, 1]) then peek_range_tombstone() will return [0, 1] before it returns [5, 6]. But reversed stream should see [5, 6] first.

Reply:
Its even more complicated than that as the native order requires tombstones to be stored like this:
[6, 5], [3, 2] and [1, 0]
When comparing we need to use `.position()` and use a native comparator. We cannot just look at the end-bound like we did earlier because of `bound_kind` is not interchangeable for end and start.

Reply:
@tgrabiec you're wrong in this - 
`_heap_cmp(s)` is initialized not with native order, but with the reversed one (the one passed to make_partition_snapshot_flat_reader`, which is the reversed schema pushed down from some other reader. On the other hand, accessing the `v.partition().row_tombstones()` needs to be done with native order, otherwise it's undefined behaviour (and indeed, causes errors on ASAN build). `_partition_schema` (or as you proposed, `_snapshot_schema`) is the native format.

In your example, `([2, 3], [5, 6])` will be selected as smaller than `([0, 1])`.

However, I acknowledge, that there is an issue with the comparator of these iterator_ranges.

Reply:
With the latest update, I (hopefully) fixed the ordering issue by changing the comparator. I need to verify if the comparator of ck positions is susceptible to this kind of error.